### PR TITLE
ignore missing blob file for iterator

### PR DESCRIPTION
**Pull request by Connor1996**
- add key-only iter option
- reproduce missing blob file when iterating
- fix an issue about sample which affecting reproducing

### CODE REVIEW

File: src/db_iter.h [hunk 1]
```diff
@@ -96,17 +109,23 @@ class TitanDBIterator : public Iterator {
       std::unique_ptr<BlobFilePrefetcher> prefetcher;
       status_ = storage_->NewPrefetcher(index.file_number, &prefetcher);
       if (status_.IsCorruption()) {
+        // If use `DeleteFilesInRanges`, we may encounter this failure,	
+        // because `DeleteFilesInRanges` may expose an old key which
```
Comment (tabokie) [inline src/db_iter.h]: "whose"

File: src/db_iter.h [hunk 2]
```diff
@@ -96,17 +109,23 @@ class TitanDBIterator : public Iterator {
       std::unique_ptr<BlobFilePrefetcher> prefetcher;
       status_ = storage_->NewPrefetcher(index.file_number, &prefetcher);
       if (status_.IsCorruption()) {
```
Comment (tabokie) [inline src/db_iter.h]: Seems vulnerable to me. Any other unexpected and unresolved corruption could cause an infinite loop. Should work as a temporary patch though.

Comment (Connor1996) [reply src/db_iter.h]: why infinite loop? Line 93 will break the loop in final

File: src/db_iter.h [hunk 3]
```diff
@@ -96,17 +109,23 @@ class TitanDBIterator : public Iterator {
...
+        // If use `DeleteFilesInRanges`, we may encounter this failure,	
+        // because `DeleteFilesInRanges` may expose an old key which	
+        // corresponding blob file has already been GCed out, so we	
+        // cannot abort here.
         fprintf(stderr,
                 "key:%s GetBlobValue err:%s with sequence number:%" PRIu64 "\n",
                 iter_->key().ToString(true).c_str(), status_.ToString().c_str(),
                 options_.snapshot->GetSequenceNumber());
+        return false;
```
Comment (yiwu-arbug) [inline src/db_iter.h]: change the `fprintf` to print to info log? And I think we shall print only once for each iterator.

File: src/db_iter.h (hunk 2 above)
Comment (tabokie) [reply src/db_iter.h]: No loop then. But we could do better than catching a generic error in future.

**Comment by Connor1996** [comment]
PTAL again @yiwu-arbug @tabokie

File: src/blob_gc_job.cc [hunk 4]
```diff
@@ -161,13 +161,15 @@ bool BlobGCJob::DoSample(const BlobFileMeta* file) {
...
 
   // TODO: add do sample count metrics
-
+  auto records_size = file->file_size() - BlobFileHeader::kEncodedLength -
+                      BlobFileFooter::kEncodedLength;
   Status s;
   uint64_t sample_size_window = static_cast<uint64_t>(
-      file->file_size() * blob_gc_->titan_cf_options().sample_file_size_ratio);
-  Random64 random64(file->file_size());
+      records_size * blob_gc_->titan_cf_options().sample_file_size_ratio);
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/blob_gc_job.cc]: for sampling, not include header and footer when calculating the discardable ratio.

**Comment by Connor1996** [comment]
PTAL again @yiwu-arbug @tabokie

File: src/blob_gc_job_test.cc [hunk 5]
```diff
@@ -84,14 +86,23 @@ class BlobGCJobTest : public testing::Test {
...
+
   void DestroyDB() {
     Status s __attribute__((__unused__)) = db_->Close();
     assert(s.ok());
     delete db_;
     db_ = nullptr;
   }
 
-  void RunGC() {
+  void RunGC(bool expected = false) {
```
Comment (tabokie) [inline src/blob_gc_job_test.cc]: Seems a lot similiar with `TEST_StartGC`. Maybe remove and update that function?

Comment (Connor1996) [reply src/blob_gc_job_test.cc]: let's just keep that in this PR. IMO `TEST_StartGC` is still too similar to `BackgroundGC`, we can merge them together.

File: include/titan/db.h [hunk 6]
```diff
@@ -77,6 +77,26 @@ class TitanDB : public StackableDB {
...
+
+  using StackableDB::NewIterators;
+  Status NewIterators(
+      const ReadOptions& options,
+      const std::vector<ColumnFamilyHandle*>& column_families,
+      std::vector<Iterator*>* iterators) override {
+    return NewIterators(TitanReadOptions(options), column_families, iterators);
+  }
+  virtual Status NewIterators(
+      const TitanReadOptions& options,
```
Comment (yiwu-arbug) [inline include/titan/db.h]: Shall we split the struct into two (e.g. `NewIterators(rocksdb::ReadOptions, TitanReadOptions, ...)`)? That's consistent with what we do on rust-rocksdb side.

File: src/db_iter.h [hunk 7]
```diff
@@ -98,14 +98,14 @@ class TitanDBIterator : public Iterator {
   }
 
   Slice value() const override {
-    assert(Valid());
+    assert(Valid() && !options_.key_only);
```
Comment (yiwu-arbug) [inline src/db_iter.h]: return `Slice()` if `key_only=true`?

Comment (Connor1996) [reply src/db_iter.h]: should we abort or set the status to `not ok`?

Comment (yiwu-arbug) [reply src/db_iter.h]: Just return `Slice()`? I think it's ok to call `value()` in case `key_only=true`.

File: include/titan/db.h (hunk 6 above)
Comment (yiwu-arbug) [reply include/titan/db.h]: discussed offline. Let's keep it the current way.

**Pull request merged**